src/Microsoft.TestPlatform.Client/Discovery/DiscoveryRequest.cs

```diff
@@ -394,54 +393,94 @@ public void HandleLogMessage(TestMessageLevel level, string message)
         /// <param name="rawMessage">Raw message.</param>
         public void HandleRawMessage(string rawMessage)
         {
-            if (this.requestData.IsTelemetryOptedIn)
+            // Note: Deserialize rawMessage only if required.
+
+            var message = this.LoggerManager.LoggersInitialized || this.requestData.IsTelemetryOptedIn ?
+                    this.dataSerializer.DeserializeMessage(rawMessage) : null;
+
+            if (string.Equals(message?.MessageType, MessageType.DiscoveryComplete))
             {
-                var message = this.dataSerializer.DeserializeMessage(rawMessage);
+                var discoveryCompletePayload = this.LoggerManager.LoggersInitialized || this.requestData.IsTelemetryOptedIn ?
+                    this.dataSerializer.DeserializePayload<DiscoveryCompletePayload>(message) : default(DiscoveryCompletePayload);
+                rawMessage = UpdateRawMessageWithTelemetryInfo(discoveryCompletePayload, message) ?? rawMessage;
```

Comment:
>UpdateRawMessageWithTelemetryInfo [](start = 29, length = 33)

what if telemetry is disabled, do we still want to add telemetry data to message?

Reply:
No we are not adding telemetry data when its disabled